lib/engine/game/g_1849/step/buy_token.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/base'
+
+module Engine
+  module Game
+    module G1849
+      module Step
+        class BuyToken < Engine::Step::Base
+          ACTIONS = %w[buy_token pass].freeze
+          MIN_PRICE = 1
+
+          def actions(entity)
+            return [] if entity != current_entity
+
+            ACTIONS
+          end
+
+          def setup
+            @bought_token = false
+          end
+
+          def can_buy_token?(entity)
+            @game.buy_tokens_enabled &&
+              !@bought_token &&
+              !available_tokens(entity).empty? &&
+              MIN_PRICE <= buying_power(entity)
+          end
+
+          def available_tokens(entity)
+            entity.tokens_by_type
+          end
+
+          def can_sell_token?(token, entity)
+            token != token.corporation.tokens.first &&
+            !token.city.tokened_by?(entity)
+          end
+
+          def any_buyable_tokens_placed?(entity)
+            @game.token_graph_for_entity(entity).connected_nodes(entity).keys.each do |node|
+              next unless node.city?
+
+              node.tokens.each do |token|
+                next unless token
+                next if token.corporation == entity
+
+                return true if can_sell_token?(token, entity)
+              end
+            end
+            false
+          end
+
+          def auto_actions(entity)
+            return [Engine::Action::Pass.new(entity)] unless any_buyable_tokens_placed?(entity)
+
+            super
+          end
+
+          # Can't buy a token from another corp if your corp already has a token on that city
+          def can_token_city?(entity, city)
+            city.tile.nodes.select(&:city?).none? { |c| c.tokened_by?(entity) }
+          end
+
+          def can_replace_token?(entity, token)
+            return false unless token
+
+            other_corporation = token.corporation
+            city = token.city
+            entity != other_corporation &&
+              token.hex.id != other_corporation.tokens.first.hex.id &&
+              other_corporation.player &&
+              token.used &&
+              city &&
+              can_token_city?(entity, city) &&
+              (city.pass? || other_corporation.placed_tokens.count { |t| !t.city.pass? } > 1) &&
+              @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
+          end
+
+          def description
+            'Buy a Token From Another Corporation'
+          end
+
+          def pass_description
+            'Skip (Buy Token)'
```

Comment:
Its pass if the player hits pass and skip if it automatically passes for them. I guess for now, to not complicate things, just use the default pass behavior and don't override it. We can enhance later if it becomes an issue for players.

Reply:
done